ntp.go

```diff
@@ -407,78 +409,57 @@ func parseTime(m *msg, recvTime ntpTime) *Response {
 
 	// Calculate values depending on other calculated values
 	r.RootDistance = rootDistance(r.RTT, r.RootDelay, r.RootDispersion)
-	r.CausalityViolation = causalityViolation(r.RTT, r.ClockOffset)
 
 	return r
 }
 
-func rtt(t1, t2, t3, t4 ntpTime) time.Duration {
-	// round trip delay time (https://tools.ietf.org/html/rfc5905#section-8)
-	//   T1 = client send time
-	//   T2 = server receive time
-	//   T3 = server reply time
-	//   T4 = client receive time
-	//
-	// RTT d:
-	//   d = (T4-T1) - (T3-T2)
-	a := t4.Time().Sub(t1.Time())
-	b := t3.Time().Sub(t2.Time())
-	return a - b
+// The following helper functions calculate additional metadata about the
+// timestamps received from an NTP server.  The timestamps returned by
+// the server are given the following variable names:
+//
+//   org = Origin Timestamp (client send time)
+//   rec = Receive Timestamp (server receive time)
+//   xmt = Transmit Timestamp (server reply time)
+//   dst = Destination Timestamp (client receive time)
+
+func rtt(prec int8, org, rec, xmt, dst ntpTime) time.Duration {
+	// round trip delay time
+	//   rtt = (dst-org) - (xmt-rec)
+	a := dst.Time().Sub(org.Time())
+	b := xmt.Time().Sub(rec.Time())
+	rtt := a - b
+	if rtt < toInterval(prec) {
+		rtt = toInterval(prec)
+	}
+	return rtt
 }
```

Comment:
IMHO, it's incorrect to truncate RTT to precision. Local clock may be precise enough and precision indicates some sort of remote server clock precision.

Reply:
This is what the reference implementation in the RFC does.

Reply:
I checked the reference implementation again and realized it is clamping RTT to the local system's clock precision, not the remote server's clock precision.  Since I don't know the local system's clock precision, I decided to clamp the value to zero instead.